In [1]:
# Competition-Solution/notebooks/trial/00_data_profiling/01_data_profiling_tweet_lengths.ipynb

---

##### <b>Imports</b>

In [2]:
import sys
from rich.console import Console

sys.path.append("../../../src")
import data_utils

console = Console()

##### <b>Loading the data</b>

In [3]:
df_call2action, df_fdgo, df_violence = data_utils.load_competition_data(data_dir="../../../data/raw/", type="trial")

Loading competition data...

Data of type 'trial' loaded successfully.

##### <b>Profiling the Length of the Tweets</b>

In [4]:
print(f"Subtask 1 (Call2Action):")
print(df_call2action["description"].str.len().describe())
print(f"Subtask 2 (DBO):")
print(df_fdgo["description"].str.len().describe())
print(f"Subtask 3 (Violence):")
print(df_violence["description"].str.len().describe())

Subtask 1 (Call2Action):
count    1051.000000
mean      151.322550
std       217.312598
min         2.000000
25%        44.000000
50%        87.000000
75%       176.500000
max      2794.000000
Name: description, dtype: float64
Subtask 2 (DBO):
count    1053.000000
mean      146.490028
std       234.521285
min         2.000000
25%        41.000000
50%        77.000000
75%       157.000000
max      3071.000000
Name: description, dtype: float64
Subtask 3 (Violence):
count    1051.000000
mean      160.262607
std       228.774794
min         1.000000
25%        48.000000
50%        91.000000
75%       191.000000
max      3071.000000
Name: description, dtype: float64


<u><p>Interpretation of Tweet Lengths (`description` column character length)</p></u>

*   **`count` (Number of Tweets with Length Calculated):**
    *   **C2A:** 1051 tweets.
    *   **DBO:** 1053 tweets.
    *   **VIO:** 1051 tweets.
    *   *Observation:* Matches the total row counts, shown already.

*   **`mean` (Average Character Length of Tweets):**
    *   **C2A:** ~151.6 characters on average.
    *   **DBO:** ~146.5 characters on average.
    *   **VIO:** ~160.3 characters on average.
    *   *Observation:* The average tweet length is broadly similar across subtasks, typically falling within old Twitter character limits (though some are clearly longer, see `max`). VIO tweets are slightly longer on average in this trial set.

*   **`std` (Standard Deviation of Tweet Character Lengths):**
    *   **C2A:** ~217.9 characters.
    *   **DBO:** ~234.5 characters.
    *   **VIO:** ~228.8 characters.
    *   *Observation:* The standard deviations are quite large relative to the means. This indicates a **wide spread in tweet lengths**. In other words, there are many short tweets but also some very long ones, pulling the standard deviation metric up.

*   **`min` (Minimum Character Length of a Tweet):**
    *   **C2A:** Shortest tweet is 2 characters long.
    *   **DBO:** Shortest tweet is 2 characters long.
    *   **VIO:** Shortest tweet is 1 character long.
    *   *Observation:* There are some extremely short tweets. Potential ground for investigation, to see if these are valid tweets or just noise.

*   **`25%` (1st Quartile - 25th Percentile):**
    *   **C2A:** 25% of tweets are 44 characters or shorter.
    *   **DBO:** 25% of tweets are 41 characters or shorter.
    *   **VIO:** 25% of tweets are 48 characters or shorter.
    *   *Observation:* A significant portion of tweets are relatively short.

*   **`50%` (Median - 2nd Quartile - 50th Percentile):**
    *   **C2A:** Half the tweets are 88 characters or shorter.
    *   **DBO:** Half the tweets are 77 characters or shorter.
    *   **VIO:** Half the tweets are 91 characters or shorter.
    *   *Observation:* The median length is consistently much lower than the mean, which is expected when the distribution is right-skewed by some very long tweets (confirmed by the high `max` values).

*   **`75%` (3rd Quartile - 75th Percentile):**
    *   **C2A:** 75% of tweets are ~177.5 characters or shorter.
    *   **DBO:** 75% of tweets are 157 characters or shorter.
    *   **VIO:** 75% of tweets are 191 characters or shorter.
    *   *Observation:* The majority of the tweets fall within the general range of typical Twitter length limits, but the range up to this percentile is still quite broad.

*   **`max` (Maximum Character Length of a Tweet):**
    *   **C2A:** Longest tweet is 2794 characters.
    *   **DBO:** Longest tweet is 3071 characters.
    *   **VIO:** Longest tweet is 3071 characters.
    *   *Observation:* There are some **exceptionally long** tweets in the dataset. This is highly unusual for typical Twitter data (even with expanded limits) and suggests these might be concatenated posts, or perhaps some data artifact. The collection time ranges from 2014 to 2016. At this time the global tweet limit was 140 characters.

<u><p>Summary & Implication of Tweet Lengths</p></u>

*   *Observation:* Tweets in the trial datasets have a median character length between 77-91 characters, but the distribution is heavily right-skewed due to the presence of a few very long texts (up to ~3000 characters). There are also some very short (1-2 character) tweets.
*   *Implication:*
    *   **Outliers:** Outliers such as very short and very long tweets need to be qualitatively examined. This includes checking if very short tweets are just "noise" and if very long tweets are valid concatenated posts or merely artifacts of some sort. The actions taken will depend on the technical approach taken to classify the tweets. (e.g. Transformer-based models, LSTM, RNNs, etc.)
    *   **Model Input Length (`max_length`):** The presence of very long texts means that a standard `max_length` for many encoder-based models (e.g., 128, 256, or 512) will lead to truncation for these outliers. This is probably acceptable, as most of our tweets fall well within these limits (e.g. 75% are under ~190 characters). The mean, median and standard deviation tweet length values that we have found out for the subtasks are going to aid in choosing a maximum input length that is both performant as well as captures the majority of the relevant data.

##### Profiling the shortest and longest tweets

In [5]:
for df, subtask_name in zip([df_call2action, df_fdgo, df_violence], ["Call2Action", "DBO", "Violence"]):
    console.print(f"Subtask: {subtask_name}", style="bold")
    shortest_tweets_series, longest_tweets_series = data_utils.get_shortest_longest_tweets(
        dataframe=df,
        column="description",
        top_n=3,
        short_length_range=3,
        long_length_range=500
    )
    
    console.line(2)

Subtask: Call2Action

Getting up to 3 tweets from specified length ranges (Short range span: 3, Long range span: 500)...

--- Displaying up to 3 Shortest Tweets (Lengths from 2 to 4) ---

Index: 111, Length: 2

Text: ":D"

Index: 207, Length: 2

Text: "Nö"

Index: 567, Length: 2

Text: "<3"

--- Displaying up to 3 Longest Tweets (Lengths from 2295 to 2794) ---

Index: 782, Length: 2794

Text: "Dresden Drecksvolk ..na wie gut.daß wir dich süßes schnckeliges kerlchen nun haben ,das gerne schmutzige 
worte in seine zuckersüße schnute nimmt..was wäre deutschland nur ohne solche bereicherer wie dir ? richtig 
:schöner :)  ...wegen solcher evolutiontragödien wie dir bin ich sowas von froh,nicht mehr im westsektor zu 
leben....man merkt gleich- du bist ein mann von welt ! ein prolet der ersten stunde !...denkst du ernsthaft,daß man
sich über deinen erguß der lächerlichkeit aufregt ? du süßer fratz,da muß ich dich enttäuschen ! du verursachst 
keine zornesfalten oder gar sorgenfalten sondern LACHFALTEN ! ...wenn ich dich kleinen wonneproppen so lese,dann 
habe ich das bild eines quengelnden kleinkindes vor mir ,das pampers und schlabberlatz trägt und flennt weil es 
seinen schnuller nicht findet ...was bin ich froh,der weichspülkultur entkommen zu sein, in der zu leider immer 
weiter zu verlottern drohst.... und je mehr ich von dir lese,desto froher bin ich,daß ich inzwischen in sachsen 
lebe und evolutionspannen wie dich nimmer so oft live+in farbe ertragen zu müßen....und wenn dich gebürtige 
ddr-bürger so lesen,dann könnte ich mir durchaus vorstellen,daß der eine oder andere sich denkt :lieber auf bananen
verzichten und dafür dich von der backe haben !  (übrigens : es gab auch ab und an bananen  in der ddr ..nicht so 
oft und nicht in rauhen mengen aber dadurch wußte man das auch zu schätzen...wie lange ist es her,daß du dich über 
so eine kleinigkeit freuen konntest ? ach,richtig..kannst du ja gar nicht..weil für dich alles selbverständlich ist
und du in deiner selbstgefälligen wessi art denkst, die weisheit mit löffeln gefuttert zu haben....merke auf :du 
hast in deinem leben absolut nix dazu beigetragen, daß deine ahnen rein zufällig in dem von den bubble-gums 
beherschten sektor gelebt haben....du bist einfach nur ein verwöhntes WOHLSTANDSNICHTS , das keinen schimmer von 
den menschen hier hat aber denkt, große reden schwingen zu können...merkst du eigentlich,wie du meine ex-landsleute
hier dahstehen läßt (ja,ich bin wessi !) mit deine schimpfwortarie ? so wie du hier den westteil der (noch)republik
repräsentierst ,vermittelst du den eindruck,daß alle wessis proleten sind....traurig ist das ! wenn du nicht in der
lage bist,deinen unmut über bautzen z.b. in manierlichem tonfall zum ausdruck zu bringen, dann hast du ein problem 
- nicht die menschen hier vor ort.....nimm es mir nicht übel,aber ich nehme dich nicht ernst...anbei :ich lebe nun 
hier vor ort und sage dir ins gesicht :lieber bin ich "dresden drecksvolk" als auch nur im entferntesten mit dir in
zusammenhang gebracht zu werden ! und daher : DRESDEN ZEIGT WIE'S GEHT ! WIR SIND DAS VOLK ! A U S M I S T E N ! 
WIDERSTAND ! MERKEL MUSS WEG ! I <3 SACHSEN ! JE SUIS SACHSE ! ..und nun :zieh leine !"

Index: 588, Length: 2389

Text: "Warum macht sich Pegida nicht für die Themen stark, die Europa in Wahrheit dominieren aber aufgrund der 
Terror-Meldungen und Hetzkampagnen der Mainstream Medien nicht wahrgenommen werden? Ihr kritisiert die Auswirkungen
der Asylpolitik statt euch mal wirklich mit den Ursachen auseinanderzusetzen. Warum haben wir so viele Flüchtlinge 
und künftig noch mehr? Die Integrationspolitik hat bislang auf ganzer Linie versagt und es werden 2015 weitere 
230000 hilfesuchende Menschen  hinzukommen. 163000 Anträge sind offen. Und die kommen mit Sicherheit nicht alle 
hierher weil es so toll bei uns ist, sondern weil sie aus ihrem Land vertrieben wurden. Wendet euch endlich den 
Ursachen zu statt die Symptome zu bekämpfen. Die Ursache ist KRIEG. Kriege, die von Deutschland als drittgrößter 
Waffenlieferant der Welt gefördert werden. Es geht bei diesen Kriegen nicht um die Menschen sondern um Macht und 
ökonomische Interessen. Früher diente die Ökonomie dem Menschen heute ist es umgekehrt. Die Menschen dienen 
weltweit einem System von dem nur Eliten profitieren. Statt ständig über Religion, Hautfarbe und Rasse zu 
diskutieren, sollten wir uns als MENSCHEN zusammenschließen und für allgemeingültige Werte einsetzen. Reduziert 
euer Denken nicht auf eine vermeintliche Islamisierung sondern auf den drohenden Verlust unserer Demokratie und 
Selbstbestimmung. Fangt  an euer Leben selbst in die Hand zu nehmen. Informiert euch bei den unabhängigen Medien. 
Unterstützt sie. Geht auf die Strasse. Setzt euch für den Frieden ein. Hier, in der Ukraine, in Syrien, im Gaza, 
weltweit. Die Angst vor einem Krieg in Europa ist groß und berechtigt. Informiert euch über TTIP, ESM, TISA. Mit 
dem heutigen Beschluss von EZB-Chef Draghi insgesamt für 1,14 B I L L I O N E N Euro Staatsanleihen anzukaufen und 
den Markt mit Geld zu fluten, für das faktisch kein Gegenwert geschaffen werden kann bewegen wir uns weiter in 
Richtung EU-Diktatur. Die Anleihen werden nie und nimmer von den jeweiligen Ländern zurückgezahlt werden. Folglich 
werden sie gemäß EU-Vertrag vergemeinschaftlicht. Was das bedeutet kann sich jeder selbst ausmalen. Wann hört 
dieser Wahnsinn endlich auf? Der Kapitalismus ist am Ende. Die Folge daraus kann nur Krieg oder Frieden sein. Ich 
schließe diesen Kommentar im Sinne eines 1. Weltfriedens mit den Worten von Bertold Brecht:
"Wenn Unrecht zu Recht wird, 
Wird Widerstand zur Pflicht""

Finished getting tweets from specified length ranges.

Subtask: DBO

Getting up to 3 tweets from specified length ranges (Short range span: 3, Long range span: 500)...

--- Displaying up to 3 Shortest Tweets (Lengths from 2 to 4) ---

Index: 130, Length: 2

Text: "??"

Index: 374, Length: 2

Text: ":D"

Index: 535, Length: 2

Text: "Nö"

--- Displaying up to 3 Longest Tweets (Lengths from 2572 to 3071) ---

Index: 968, Length: 3071

Text: "Flüchtlinge, Flüchtlinge, Flüchtlinge: Wenn ich das mit 1945 vergleiche, wird mir schlecht. Die meisten 
Deutschen sind in Deutschland geblieben, obwohl der Krieg tobte und alles zerbombt war. Es bestand Hungersnot, die 
Russen vergewaltigten die Frauen. Aber keiner konnte abertausende Kilometer durch sichere Drittstaaten spazieren, 
um ins gelobte Hartz4-Land zu kommen. Nahrungsangebote hätte wohl damals keiner zurückgewiesen. Für jedes Stückchen
Brot waren sie dankbar. Und sie blieben im Elend, halfen beim Wiederaufbau. Die Bilder von Trümmerfrauen kennt 
jeder. Und heute kommen junge Männer mit aggressiven Glauben und Merkel-Merkel-Transparenten. Es ist die Mutti mit 
den Geldgeschenken und mit der Hoffnung auf leichten Wohlstand. Für diese Absicherung ist man auch bereit, 
Schleuser zu bezahlen oder sich falsche Pässe zu besorgen, um die wahre Identität zu verbergen. Überall hat es sich
in der arabischen Welt heraumgesprochen. Jeder darf kommen, die Selfies und Smartphones verkünden es. Aber 
natürlich gibt es nun auch Opfer. Es werden Deutsche sein. Ob Rentner, ob Arbeitslose, ob Minijobber, ob 
Zeitarbeiter oder Mindestlöhner. Der Verteilungskampf beginnt. Um Sozialleistungen, um Bildungschancen, um 
bezahlbaren Wohnraum. Ich will mir gar nicht vorstellen, wenn die Hand, die derzeit füttert, mal nicht mehr füttern
kann oder will. Wenn Sozialleistungen gestrichen werden, wenn Rückführungen anstehen, weil die Heimat befriedet ist
- was dann passiert. Dann sehen wir das echte Aggressionspotenzial. Jeder kennt die Bilder mit den Massen wütender 
Lynchmobs, die Fahnen verbrennen, weil einer in Europa eine Karrikatur malt. Deutschland ist naiv, es steht völlig 
allein, trägt freiwillig alle Kosten und Risiken, handelt genauso überstürzt, wie bei der Energiewende. Bluten muss
wie immer der kleine Mann auf der Straße, die Melkkuh, der keiner Hilft, wenn sie überfallen, verprügelt oder 
bestohlen wird. Wie Integration aussieht, kann man in Justizvollzugsanstalten sehen. Oder in Duisburg. Einen 
kleinen Vorgeschmack gab es schon - Suhl. Wieviele Europäer wären wohl zum Lynchen übergegangen, weil jemand ein 
paar Seiten aus der Bibel reisst? Richtig, keiner. Junge Männer bedeuten Gewalt, bedeuten aber auch sexuelle 
Bedürfnisse - alles in anbetracht, dass Frauen keinerlei Respekt bei ihnen genießen. Konflikte von dort, kommen 
hier her. Es sind ihre Brüder und Schwestern, die sich dort die Köpfe einschlagen, die grausam sind - es sind ihre 
Landsleute. Mit der selben Erziehung, mit der selben Wertevorstellung, mit dem selben Glauben, womöglich mit dem 
selben Charakter. Es sind keine höflichen und fleißigen Thailänder. Nein, es sind laute und fordernde Menschen, die
jederzeit zum heiligen Krieg bereit sind. Ihre Integration wird scheitern. Ihre Bildung ist gering. Sprachbarrieren
führen zu Perspektivlosigkeit, zu relativer Armut. Ihre Hemmschwelle ist gering. Es wird Stadtviertel geben, die 
Deutsche und "Gottlose" nicht mehr betreten können.Und irgendwann wird sich das alles entladen, natürlich nicht im 
Kanzleramt. Gute Nacht Deutschland!"

Finished getting tweets from specified length ranges.

Subtask: Violence

Getting up to 3 tweets from specified length ranges (Short range span: 3, Long range span: 500)...

--- Displaying up to 3 Shortest Tweets (Lengths from 1 to 3) ---

Index: 57, Length: 1

Text: "?"

Index: 202, Length: 2

Text: "??"

Index: 310, Length: 2

Text: "-1"

--- Displaying up to 3 Longest Tweets (Lengths from 2572 to 3071) ---

Index: 781, Length: 3071

Text: "Flüchtlinge, Flüchtlinge, Flüchtlinge: Wenn ich das mit 1945 vergleiche, wird mir schlecht. Die meisten 
Deutschen sind in Deutschland geblieben, obwohl der Krieg tobte und alles zerbombt war. Es bestand Hungersnot, die 
Russen vergewaltigten die Frauen. Aber keiner konnte abertausende Kilometer durch sichere Drittstaaten spazieren, 
um ins gelobte Hartz4-Land zu kommen. Nahrungsangebote hätte wohl damals keiner zurückgewiesen. Für jedes Stückchen
Brot waren sie dankbar. Und sie blieben im Elend, halfen beim Wiederaufbau. Die Bilder von Trümmerfrauen kennt 
jeder. Und heute kommen junge Männer mit aggressiven Glauben und Merkel-Merkel-Transparenten. Es ist die Mutti mit 
den Geldgeschenken und mit der Hoffnung auf leichten Wohlstand. Für diese Absicherung ist man auch bereit, 
Schleuser zu bezahlen oder sich falsche Pässe zu besorgen, um die wahre Identität zu verbergen. Überall hat es sich
in der arabischen Welt heraumgesprochen. Jeder darf kommen, die Selfies und Smartphones verkünden es. Aber 
natürlich gibt es nun auch Opfer. Es werden Deutsche sein. Ob Rentner, ob Arbeitslose, ob Minijobber, ob 
Zeitarbeiter oder Mindestlöhner. Der Verteilungskampf beginnt. Um Sozialleistungen, um Bildungschancen, um 
bezahlbaren Wohnraum. Ich will mir gar nicht vorstellen, wenn die Hand, die derzeit füttert, mal nicht mehr füttern
kann oder will. Wenn Sozialleistungen gestrichen werden, wenn Rückführungen anstehen, weil die Heimat befriedet ist
- was dann passiert. Dann sehen wir das echte Aggressionspotenzial. Jeder kennt die Bilder mit den Massen wütender 
Lynchmobs, die Fahnen verbrennen, weil einer in Europa eine Karrikatur malt. Deutschland ist naiv, es steht völlig 
allein, trägt freiwillig alle Kosten und Risiken, handelt genauso überstürzt, wie bei der Energiewende. Bluten muss
wie immer der kleine Mann auf der Straße, die Melkkuh, der keiner Hilft, wenn sie überfallen, verprügelt oder 
bestohlen wird. Wie Integration aussieht, kann man in Justizvollzugsanstalten sehen. Oder in Duisburg. Einen 
kleinen Vorgeschmack gab es schon - Suhl. Wieviele Europäer wären wohl zum Lynchen übergegangen, weil jemand ein 
paar Seiten aus der Bibel reisst? Richtig, keiner. Junge Männer bedeuten Gewalt, bedeuten aber auch sexuelle 
Bedürfnisse - alles in anbetracht, dass Frauen keinerlei Respekt bei ihnen genießen. Konflikte von dort, kommen 
hier her. Es sind ihre Brüder und Schwestern, die sich dort die Köpfe einschlagen, die grausam sind - es sind ihre 
Landsleute. Mit der selben Erziehung, mit der selben Wertevorstellung, mit dem selben Glauben, womöglich mit dem 
selben Charakter. Es sind keine höflichen und fleißigen Thailänder. Nein, es sind laute und fordernde Menschen, die
jederzeit zum heiligen Krieg bereit sind. Ihre Integration wird scheitern. Ihre Bildung ist gering. Sprachbarrieren
führen zu Perspektivlosigkeit, zu relativer Armut. Ihre Hemmschwelle ist gering. Es wird Stadtviertel geben, die 
Deutsche und "Gottlose" nicht mehr betreten können.Und irgendwann wird sich das alles entladen, natürlich nicht im 
Kanzleramt. Gute Nacht Deutschland!"

Index: 929, Length: 2851

Text: "Sorgen und Befürchtungen haben wir alle, Herr Grüner. Links genauso wie Rechts. Aber es macht doch keinen 
Sinn, in Wahnsinn zu verfallen und allen Böses zu unterstellen. Ich habe nicht genug Ahnung und auch keinen 
Kristall, der mit die Zukunft deutet, um zu wissen, was wirklich aufgrund der Flüchtlinge, aufgrund des Lobbyismus,
aufgrund der gnadenlosen Ausnutzung unseres Planeten, aufgrund der Raffgier passieren wird. Aber ich weiß, dass ich
niemals meine Menschlichkeit und mein Mitgefühl verlieren möchte. Und ich fühle wirklich mit jedem mit (ich leide 
nicht mit - ich fühle mit!), der von Terror verfolgt wird, der um die Zukunft seiner Kinder bangt, der nicht weiß, 
wie es morgen weitergehen soll und auch dem, der "einfach nur" Hunger leidet. So wie ich auch Ihre Ängste, ihre Wut
wahrnehme.
Und ich weiß, dass alle Länder dieser Welt viel näher aneinander sind als jemals zuvor. Dank Filmen, dank Internet 
etc etc etc. Und ich weiß, dass unser Lebensstil sehr attraktiv wirkt. So wie damals auch die BRD sehr attraktiv 
auf die DDR-Bürger wirkte. Den oftmals so güldenen Bildern des West-Fernsehens und seiner Werbespots und Serien sei
dabei auch Dank. Ich weiß auch, dass IS unendlich grausam ist und ALLES und ALLE (auch Muslime!), die sich nicht zu
ihnen bekehren, auf teilweise schrecklichste Art abschlachten. Ich weiß, dass sie den Westen überrennen und mit 
Terror überfluten wollen. Es ist mir alles bewusst. Und ich habe sogar Mitgefühl mit ihnen. Denn wer solch einen 
Hass hat, dem muss es eigentlich unendlich Scheiße gehen. (was nicht heißt, dass ich sie auch nur ansatzweise mag).
Natürlich weiß ich auch, dass wir nicht alle Flüchtlinge aufnehmen können (und ich mag jetzt gar nicht diskutieren,
dass damit auch IS zu uns kommen könnte). Nur wenn wir alle überlegen, wie man das Leid dort beenden kann, wo es 
entsteht (und das muss auch heißen auf Teile unseres Luxus - ja, wir alle leben immer noch im Luxus! -  zu 
verzichten, da die Welt nicht allen Menschen Autos und Fleisch und und und geben kann), haben wir eine Chance auf 
eine Welt, in der wir alle hoffentlich irgendwann mal gut leben können. 
Und nein, Politikerin will ich wirklich nicht sein :D a) ist meine Haut dafür zu dünn b) würde ich mich mit Hinz 
und Kunz streiten müsse, c) mich ständig fragen würde, ob ich vielleicht auf dem komplett falschen Dampfer mit 
meinen Gedanken bin, d) ich einfach keine Ahnung habe, wie ein gangbarer Weg zu einer wirklich guten Lösung 
aussehen kann und e) würden Sie wahrscheinlich ganz schnell auf mich schimpfen, weil ich Dinge vorschlagen würde, 
die nicht angenehm sind. Denn wie soll bitte ein friedliches Leben funktionieren, ohne dass wir alle uns ändern. 
Und ja, ich bin Utopistin und sicherlich auch Gutmensch - aber ich wünsche mir wirklich, dass irgendwann mehr 
Frieden und vor allem Glück auf dieser Welt herrscht."

Finished getting tweets from specified length ranges.